### DISCUSSION TITLE
How to initialise a ValueNotifier (useState) to a Riverpod future? in a HookConsumerWidget

### DISCUSSION THREAD

**Discussion by linxydmhg** [rrousselGit/riverpod#2135] [Q&A]
Hello, I am facing a small problem..

I want to initialise a useState inside my HookConsumerWidget with a value from a riverpod future, but I can't seem to figure out how exactly. I have working code but it just doesn't look right at all..

```
   class SomeHookConsumerWidget {
   ...
  @override
  Widget build(BuildContext context, WidgetRef ref){

   // These are the ValueNotifiers that I want to initialize to a riverpod future value
    final ValueNotifier<Option<int>> selectedCalibre = useState(const None());
    final ValueNotifier<Option<FieldLot>> selectedLot = useState(const None());

    // here I use ref.read inside a useEffect
    useEffect(() {
      ref.read(palletProvider(palletId)).whenData((value) => value.match(() {}, (pallet) {
            int? initialCalibre = pallet.calibreBoxes.isNotEmpty ? pallet.calibreBoxes.keys.first : null;
            if (initialCalibre != null) {
              selectedCalibre.value = Some(initialCalibre);
            }
          }));

      return null;
    }, []);

    // here I try to setup a listener to initialize inside the useEffect
    useEffect(
      () {
        final sub = ref.listenManual(plantProvider, (previous, next) {
          if (next.hasValue) {
            print('next.value is ${next.value}');
            selectedLot.value = Some(next.value!.currentFieldLot);
          }
        });
        return sub.close;
      },
      [],
    );
```

Both ways ive used look like they are not the way to do it. Can anyone give me some guidance on how I can initialize those useState with a riverpod future value?

**Comment by AhmedLSayed9** [comment]
`ref.read(palletProvider(palletId)).whenData` will only trigger one time and supposing your future might has no data yet, your whenData closure will not be called.

Instead, you can register a listener that will be called whenever new state is emitted, which is what you're doing in your second useEffect but you need to set `fireImmediately` to true in case your provider is being listened to in somewhere else and its future already completed.

Also, you don't need to close the subscription as riverpod is doing that for you when the widget is disposed.

It should be like this:
```
useEffect(() {
  ref.listenManual(
    palletProvider(palletId),
    (previous, next) {
      //Do your thing....
    },
    fireImmediately: true,
  );
  ref.listenManual(
    plantProvider,
    (previous, next) {
      /Do your thing....
    },
    fireImmediately: true,
  );
  return null;
}, []);
```
